Tharun/Assignment_4a.cu

```diff
@@ -0,0 +1,139 @@
+#include<iostream>
+using namespace std;
+int Array_Size;
+
+__global__ void Min(float* d_in1, float* d_out,int* d_array_size)
+{
+    int id = blockIdx.x * blockDim.x + threadIdx.x;
+	int t_id = threadIdx.x;
+    int b_id = blockIdx.x;
+	
+	__shared__ float a[1024];
+	
+    if(id < *d_array_size)
+     a[t_id] = d_in1[id];    
+	
+	__syncthreads();
+	
+	for(int s = 512; s>0; s = s/2)
+    {
+        __syncthreads();
+        if(id>=*d_array_size || id+s>=*d_array_size)
+            continue;
+        if(t_id<s)
+            {
+               if(a[t_id] > a[t_id + s])
+                a[t_id]= a[t_id + s];
+            }
+        // Each iteration reduces size of active array by half
+    }
+    __syncthreads();
+	
+	 if(t_id==0)
+        d_out[b_id] = a[t_id];   
+}
+__global__ void Max(float* d_in1, float* d_out,int* d_array_size)
+{
+    int id = blockIdx.x * blockDim.x + threadIdx.x;
+	int t_id = threadIdx.x;
+    int b_id = blockIdx.x;
+	
+	__shared__ float a[1024];
+	
+    if(id < *d_array_size)
+     a[t_id] = d_in1[id];    
+	
+	__syncthreads();
+	
+	for(int s = 512; s>0; s = s/2)
+    {
+        __syncthreads();
+        if(id>=*d_array_size || id+s>=*d_array_size)
+            continue;
+        if(t_id<s)
+            {
+               if(a[t_id] < a[t_id + s])
+                a[t_id] = a[t_id + s];
+            }
+        // Each iteration reduces size of active array by half
+    }
+    __syncthreads();
+	
+	 if(t_id==0)
+        d_out[b_id] = a[t_id];   
+}
+float Find_min_CPU(float h_array[], int Size)
+{
+    float naive_min = h_array[0] ;
+    for(int i=0; i<Size-1; i++)
+         {
+            if(h_array[i]>h_array[i+1])
+            naive_min=h_array[i+1];
+         }
+    return naive_min;
+}
+
+
+
+float Find_max_CPU(float h_array[], int Size)
+{
+    float naive_max = h_array[0];
+    for(int i=0; i<Size-1; i++)
+         {
+            if(h_array[i]<h_array[i+1])
+            naive_max=h_array[i+1];
+         }
+    return naive_max;
+}
```

Comment:
Your serial code for calculating min and max is wrong. 
Suppose the array is {10, 4, 6}, your Find_max_CPU function returns 6, and not 10.